### ISSUE TITLE
How to scale down a training job from 8 GPUs to 4 GPUs while keeping world size same

### ISSUE THREAD

**Issue by botieking98** [intelligent-machine-learning/dlrover#804]
I find that dlrover can only scale up/down the worker number of training job elasticly.

Assume a training job runs on **a single node with 8 GPUs**(one worker process per GPU)，now I want to scale down the training job from 8 GPUs to 4 GPUs，i.e **8 worker process running on 4 GPUs**(GPU time-slicing)，worker number of the job should keep same during scale up/down.

Is there any way to achieve this?

**Comment by workingloong** [comment]
Do you want to keep the batch size fixed using `8 worker process running on 4 GPUs`? If you want to keep the batch size fixed when the worker process scales down, gradient accumulation may be a better choice. Because, the CUDA memory will double if we run 2 processes on one GPU and OOM may happen. We have implemented the feature to automaticaly set `gradient_accumulation_step` according to the number of process if the total number `WORKER_NUM` is fixed. The detail is in 
https://github.com/intelligent-machine-learning/dlrover/blob/master/dlrover/trainer/torch/elastic/trainer.py#L420-L440

If you want to scale down GPUs on one node, you may need to set `CUDA_VISIBLE_DEVICES=0,1,2,3` and restart the training processes. 

One question: When do we need to scale a job from 8 GPUs to 4 GPUs on a single node?

**Comment by botieking98** [comment]
Generally, I do not consider global batch size is or not fixed, I want to implement **preemptive and elastic** DNN training. 
1) When GPU resources are sufficient, I need to increase the number of workers to fully use GPU resources to accelerate the training process. 
2) When GPU resources are insufficient or **some inference tasks need to be executed priorily**, part of GPU resources need to be released for inference tasks. 
However, I do not want to remove some training workers and need to run all workers on the remaining GPUs. For users, the number of workers does not change, but the GPU resources that can be used are reduced.

I have found `Singularity` is a good solution, but that is seemly not open source.
[Singularity: Planet-Scale, Preemptive and Elastic Scheduling of AI Workloads](https://arxiv.org/pdf/2202.07848.pdf)

**Comment by workingloong** [comment]
It is very difficult to not restart the training process to support all DDP, FSDP, DeepSpeed Zero and Megatron when the number of training GPUs changes. 
DLRover can support  the  preemptive and elastic scheduling scenario on a k8s cluster. The training job has some high-priority Pods and some low-priority Pods. Low-priority Pods can be preempted by high-priority Pods of the inference job on a k8s cluster. After the inference job releases some Pods, the training job can relaunch the low-priority Pods.

Singularity is good solution for DDP but not for FSDP, DeepSpeed Zero and Megatron. The later distributed training strategies partition the model and optimizer across all GPUs to train a large fundational model. For example, one process need to gather model and optimizer parameters from other process in FSDP. I don't know how to implement it by time-slicing. What's more, OOM may happen if the number of GPUs is small.

**Comment by botieking98** [comment]
> It is very difficult to not restart the training process to support all DDP, FSDP, DeepSpeed Zero and Megatron when the number of training GPUs changes. DLRover can support the preemptive and elastic scheduling scenario on a k8s cluster. The training job has some high-priority Pods and some low-priority Pods. Low-priority Pods can be preempted by high-priority Pods of the inference job on a k8s cluster. After the inference job releases some Pods, the training job can relaunch the low-priority Pods.
> 
> Singularity is good solution for DDP but not for FSDP, DeepSpeed Zero and Megatron. The later distributed training strategies partition the model and optimizer across all GPUs to train a large fundational model. For example, in a training step of FSDP, one process need to gather model and optimizer parameters from other process in the forward and scatter all gradients to other processes in the backward. I don't know how to implement it by time-slicing. What's more, OOM may happen if the number of GPUs is small.

em... let me think about a better way to do it. 
Thanks for your reply.

**Comment by TomSuen** [comment]
> It is very difficult to not restart the training process to support all DDP, FSDP, DeepSpeed Zero and Megatron when the number of training GPUs changes....

Hi, how does dlrover used with DeepSpeed Zero? Just modify the training py code?
